### ISSUE TITLE
Idea: built-in support for promises

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx#1370]
This proposal aims to making working with promises easier.

Some ideas:

1. Expose `observable.promise(Promise)` than returns `{ state, value, promise }`. In other words, [`fromPromise`](https://github.com/mobxjs/mobx-utils#frompromise) of the mobx-utils package
2. `observable(Promise)` and `@observable x = Promise` will automatically convert to a promise observable (e.g. call `observable.promise`)
3 Allow `@computed` to return Promises, and if so, convert this again in a promise structure. 

Example code:

```javascript
class User {
    @observable name

    @computed
    get profile() {
      // return a promise
      return fetch(`/endpoint/${name}`)
    }

    @computed
    get profileImageUrl() {
       // pluck the image from the profile promise
       return this.profile.state === "fullfilled"
          ? this.profile.value.image
          : "/defaultavatar.png"
    }

    @computed({ initialValue: [] })
    get tags() {
       // chain another promise onto the profile promise
       return this.profile.then(() =>
          fetch(`/endpoint/${name}/tags`)
       })
    }

    @computed
    get isLoading() {
       // are we fetching data?
       return this.profile.state === "pending" || this.tags.state === "pending"
    }

    @computed
    get isLoadingTakingLong() {
      // provide spinner if loading takes long!
      return Promise.race([
         new Promise(resolve => setTimeout(() => resolve(true), 1000),
         new Promise(resolve => this.isLoading.then((done) => { if (done) resolve(false) })
      ])
    }
}

const UserView = observer(({ user }) => (
   <div>
       <h1>{user.name}<h1/>
       { user.isLoadingTakingLong.value && <Spinner /> }
       <img src={user.profileImageUrl} />
       { user.tags.value.join(" - ") }
   </div>
))
```

I checked, this will even work correct with debug tools like `trace`. 

## Problems

The problem with the decorators is that their return type signature is incorrect; as the type `Promise<T>` would be converted into `Promise<T> & { state: "pending" | "fullfilled" | "error", value: T | undefined | error }`

Possible solutions:

```javascript
@computed
 get profile() {
      // upcast
      return fetch(`/endpoint/${name}`) as IObservablePromise
}

@computed
get profile() {
      // wrapper fn (needing a better name)
      return observablify(`/endpoint/${name}`) 
}

// ... but still:
@computed
 get profile() {
      // upcast
      const p fetch(`/endpoint/${name}`) as IObservablePromise
      p.state // undefined!
      return p
}
```

Caveat: only the first part of a computed will be tracked. But that is explainable and makes it somewhere easy as well.

cc @spion  @urugator  @danielearwicker

**Comment by mweststrate** [comment]
cc @d4rky-pl

**Comment by urugator** [comment]
I don't mind "enhancing" Promise with observable fields. What seems like a bigger problem is the inability to control conversion of the value itself.
On `@observable` you can get the control back into your hands with ref modifier (btw should shallow convert promise, but not the value?), however that's not the case with computed.
Also this is the only case where computed converts the returned value right? That may be unexpected...
So I think it's quite reasonable on `@observable`, but in case of `@computed` I would leave the conversion to user.

Btw shouldn't it be `@computed.options({ initialValue: [] })` in your example? :)

**Comment by mweststrate** [comment]
Someone suggested `@computed.promise` to make things explicit.

Also `@observable.ref` should leave all promises as is (likewise, shallow on collections of promises).

Other concern, one might be tempted to return model classes from computed promises, but I think that might be trick and could make things very non transparent? E.g.:

```javascript
class TodoStore {
  @computed
  get todoData() {
      return fetch(`myapp/todos`)
  }

  @cmputed
  get todos() {
     if (this.todoData.state !== "fullfilled")
        return []
     return this.todoData.value.map(json => new Todo(json))
  }
}

class Todo {
  @observable title
  @observable done
}
```

**Comment by urugator** [comment]
There is no need or benefit from pushing this to own decorator or decorator options
```javascript
@computed.promise({ initialValue : "", deep: false })
get promise() {
  return fetch(`myapp/todos`);
}
// following is imho much cleaner and easier to understand
@computed
get promise() {
  const promise = fetch(`myapp/todos`)
  return observable.promise(promise, { initialValue: "" , deep: false });
}
```

**Comment by mweststrate** [comment]
@benjamingr  cc

**Comment by urugator** [comment]
> (likewise, shallow on collections of promises).

What I meant was:
```javascript
@observable.shallow = promise; 
// should work like 
@observable = observable.promise(promise, { deep: false }); // note the deep option 
```

**Comment by spion** [comment]
My idea is to not enhance promises, but instead, store a global WeakMap of PromiseStates.

```typescript
function caseAsync(p:Promise, handlers: Handlers) {
  let promiseState = weakMap.get(p);
  if (promiseState == null) { weakMap.set(p, promiseState = new PromiseState(p); }
  return promiseState.case(handlers);
}
```

Then you can use this on any promise or thenable.

**Comment by benjamingr** [comment]
We need async iterator support, not promise support.

**Comment by spion** [comment]
@mweststrate the weakmap approach I described above would play together nicely (and type safe) with some really cool modifiers. Two that I think would be super useful and gamechangers (but need deep access to mobx internals) are:

`@computed.options({debounce: N})`

When the computed becomes dirty, wait N miliseconds of idle time before recomputing and propagating dirtyness to own observers (return previous values in the meantime)

`@computed.options({replaceWhenResolved: true})`

If the computed returns a promise, don't replace the previous state until the replacement is actually resolved somehow.

**Comment by spion** [comment]
You can use replaceWhenResolved to do react style async loading:

```
class LateLoader(fetcher, delay) {
  @computed get promise() { return fetcher(); }
  @computed.options({replaceWhenResolved: true}) 
  get lateValue() {
    return Promise.race(timeout(delay), this.promise())
      .then(value => ({value}), e => {
         if (e.type !== 'timeout-error') throw e;
         else return {timeout: true}
      });
  }

  case(handlers) { 
    return mobx.caseAsync(this.lateValue, {
      fulfilled: res => res.value ? handlers.fulfilled(res.value): handlers.pending();
      rejected: e => handlers.rejected(e)
      pending: () => handlers.pending() // only initial load activates this
    })
  }
})
```

**Comment by benjamingr** [comment]
Why would we be going _against_ the grain here by adding a relatively big feature - why not do it async iterators @spion ?

With the more standards approach this would just be:

```js
class LateLoader(fetcher, delay) {
   @computed async function* lateValue() {
      while(true) { // remove to only fetch once rather than continously
        try { 
          yield await Promise.race([fetcher(), timeout(delay)]);
        } catch (e) {} // don't replace anything, or rethrow if timeout error
      }
    }
}
```

**Comment by benjamingr** [comment]
Also 

> When the computed becomes dirty, wait N miliseconds of idle time (no new "dirty" signals) before recomputing and propagating dirtyness to own observers (return previous values in the meantime)

One of the good things about mobx is the "always synchronous, with a fresh trace" approach. I don't think debounce should be the responsibility of the computed observable itself.

**Comment by spion** [comment]
@benjamingr ok, show me debounce that waits for the input field to stop changing before sending a request, with an async generator.

**Comment by benjamingr** [comment]
@spion Sure, I gave a talk about it 2 years ago when it was still a proposal, here is my old Rx autocomplete example from my ng-conf-il talk converted to async generators: https://docs.google.com/presentation/d/1r2V1sLG8JSSk8txiLh4wfTkom-BoOsk52FgPBy8o3RM/edit#slide=id.p  It's also referenced in Axel's new book about ES2018.

Doing this with mobx should be significantly easier, since instead of the ugly fromEvent with promises you'd just convert an observable to an async generator _directly_ with `Symbol.asyncIterator`

So basically:

```js
async function request() {
  let timer = null;
  for await(const value of someMobxObservable) {
    clearTimeout(timer);
    timer = setTimeout(() => sendRequest(value).then(react), delay);
  }
}
```

**Comment by spion** [comment]
@benjamingr so that breaks "transparent reactive programming" - the observable now has to have a different type, as well as implement the async iterator protocol.

I think its too early for that big change, and that in the meantime debounce is both easy to use and easy to understand.

**Comment by benjamingr** [comment]
> @benjamingr so that breaks "transparent reactive programming" - the observable now has to have a different type, as well as implement the async iterator protocol.

Observables implement "the async iterator protocol" since they are conceptually the same, just like ES observables will implement the async iteration protocol - it is only _logical_ although the for...await could easily "be done in userland". 

It's not a big implementation and I think it would be a good time - we recently landed support for conversion in Node 10, Chrome and Firefox support is here and all the transpilers support it. As platforms all converge on support I wouldn't want MobX to be left behind.

> and that in the meantime debounce is both easy to use and easy to understand.

I think that you just threw a very specific use case at me and I showed how a lower level more powerful primitive can eat it for breakfast :D If I recall correctly _you_ are the one always pushing for simpler primitives rather than a kitchen sink of examples.

Honestly I think the MobX API surface can be reduced dramatically for the average user, and this is the one thing I think would _generalize_ mobx (@computed taking an async iterator input).

**Comment by spion** [comment]
> I think that you just threw a very specific use case at me and I showed how a lower level more powerful primitive can eat it for breakfast :D If I recall correctly you are the one always pushing for simpler primitives rather than a kitchen sink of examples.

No, the problem was that you broke all assumptions underlying mobx decorator types to implement a pet theory about async generators :grinning: - mainly the fact that a mobx observable or  "computed" value is of type T, not AsyncStream<T>, and TypeScript doesn't allow a decorator to change a type.

p.s. have you checked mobx-utils? there is already `mobxUtils.toStream` in there, which you would need for your code.

**Comment by spion** [comment]
@benjamingr I would push for primitives, if promises were not so badly designed! :smiling_imp: If for example they were cancellable, we could just have "@computed" and automatically cancel the previous promise when a new one is generated. Then debouncing would be as easy as `delay(N).then(doRequest)`

**Comment by benjamingr** [comment]
`toStream` is for an observable (and not for an async iterator), observables are stage 1 whereas async iterators have already landed. My "pet theory" about async generators is literally the ECMAScript specification at its latest standard as well as how Node APIs will look in the future.

All this is _your_ vision (promise streams), now they're native, we can't run from the fact they're there and know what? They're pretty awesome :) 

A mobx observable changing over time _is_ a stream (that's basically what the Symbol.observable (which, fun fact - is changing by the way to Symbol.observe "does" in "toStream"). We can convert that stream to the one JavaScript primitive already implemented in browsers that represents just that and provide a native declarative way to process streams.

------

> If for example they were cancellable, we could just have "@computed" and automatically cancel the previous promise when a new one is generated. Then debouncing would be as easy as delay(N).then(doRequest)

Given we both vouched for cancellable promises (I might still sneak them in the platform) then we both agree. That said async iterators are iterators and are by definition a lazy primitive which cancellation works for. 

I recommend that you play with the primitive and formulate an opinion, (remember when I told you this about MobX?) - It's quite great.

**Comment by spion** [comment]
@benjamingr we need `toAsyncIterator` then. And it has to take an expression.

For example in your LateLoader code, the generator will keep fetching over and over whereas its only suppoesd to fetch when the dependencies of `fetcher()` change.

Or rather, fetcher would be `toAsyncIterator(() => fetchOnce())`

**Comment by benjamingr** [comment]
@spion I think we get to shape the future of how mobx looks now, and this works everywhere - note that implementing it is only adding a symbol so it only increases the API surface in a way that supports an existing API 

> For example in your LateLoader code, the generator will keep fetching over and over whereas its only supposed to fetch when the dependencies of fetcher() change.

That's fixable though, like you said :)

**Comment by spion** [comment]
@benjamingr I'm also pretty sure that I'd still like to avoid the mental shift of thinking about mobx as automatically refreshing (with auto dependency tracking) values to thinking about them as async iterables...

**Comment by spion** [comment]
@mweststrate yep. Except for one thing: debounce. 

In the absence of promise cancellation, we can't fire a promise immediately when the dependencies change, because that will definitely end up sending a request. 

OTOH I think replaceWhenResolved is probably doable without any additions to core.

**Comment by mweststrate** [comment]
The async iterator starts to click @benjamingr, thanks for pointing out (I admit, again ;-))

**Comment by mweststrate** [comment]
@benjamingr just checking if I understand everything correctly, we could already support `[Symbol.asyncIterator]` like this? To make every observable an async iterator. After that we only need an utility that consumes async iterators and notifies it's atom. That should be pretty easy.

```javascript
ObservableValue.prototype.[Symbol.asyncIterator] = function( {
   return {
       next() {
          return new Promise(resolve => {
               const disposer = reaction(
                   () => this.get(), 
                   (value) => {
                       disposer()
                       resolve(value)
                   }
           })
       }
   }
})
```

Obviously this iterator never ends, but I guess that makes sense, or am I confusing two paradigms?

**Comment by benjamingr** [comment]
> @benjamingr just checking if I understand everything correctly, we could already support [Symbol.asyncIterator] like this? 

Close, `.next()` can be called multiple times (so we'd need to queue results) and we'd need to buffer results while it's not iterated.

It should also probably end when `.return` is called on it - for example when `break`ing from a loop.

As Gorgi said, it's also `{value, done: false}` :)

> After that we only need an utility that consumes async iterators and notifies it's atom. That should be pretty easy.

Yes, that's just a `for await` loop over the async iterator that updates the atom. The powerful part here is the async generator can `yield` multiple changing values.

Note that since `@computed` and `@observable` don't have interesting behavior for async iterators at the moment they could probably just be overloaded to support them.

**Comment by mweststrate** [comment]
> Close, .next() can be called multiple times (so we'd need to queue results)

Yeah this is what I didn't really understand, why is this behavior needed? From technical perspective, this means simply storing the last distributed promise and chaining onto that one?

> and we'd need to buffer results while it's not iterated.
Why? We only need to expose the *next* value that becomes available, right? So until the observable changes, that first promises just remains pending:

> As Gorgi said, it's also {value, done: false} :)
Doh

Improved:

```
ObservableValue.prototype.[Symbol.asyncIterator] = function() {
   let lastPromise = Promise.resolve()
   return {
       next() {
          return lastPromise = lastPromise.then(() => new Promise(resolve => {
               const disposer = reaction(
                   () => this.get(), 
                   (value) => {
                       disposer()
                       resolve({ value, done: false })
                   }
               )
           }))
       }
   }
})

**Comment by benjamingr** [comment]
> Yeah this is what I didn't really understand, why is this behavior needed? From technical perspective, this means simply storing the last distributed promise and chaining onto that one?

Yes, see this discussion from when we implemented it in Node - https://github.com/nodejs/node/pull/17755

**Comment by benjamingr** [comment]
```js
ObservableValue.prototype.[Symbol.asyncIterator] = function() {
   let lastPromise = Promise.resolve()
   return {
       disposer: () => {}, 
       next() {
          return lastPromise = lastPromise.then(() => new Promise(resolve => {
               this.disposer = reaction(
                   () => this.get(), 
                   (value) => {
                       disposer()
                       resolve({ value, done: false })
                   }
               )
           }))
       },
       return() { 
         this.disposer();
         return ({value: undefined, done: true });
       }
   }
})
```

Although, should we keep a reference to the reactions (so we don't "miss" any if no one called `next`)?

That is, if an observable changes in this way:

```js
1 -> 2 -> 3 -> 4 -> 5
```

But we only call `.next()` manually when it's 2 and 4, do we expect to get the values 3, 4 or 3, 5?

**Comment by mweststrate** [comment]
> But we only call .next() manually when it's 2 and 4, do we expect to get the values 3, 4 or 3, 5?

I would tend to say no; MobX is not about event streams but value streams, where only the last one matters (aka. the current state)? 

On the other hand, the above setup could course observables to become temporarily unobserved, as long as `next()` is not called on the generator. So maybe the following setup is better:

```javascript
ObservableValue.prototype.[Symbol.asyncIterator] = function() {
   let lastPromise = Promise.resolve()
   let resolver
   let disposer = this.observe(v => {  // observe is more efficient then reaction, but further the same
       if (resolver) resolver(v)
       resolver = null 
       // or, queue here... to make sure *all* values get propagated
   }) 
   return {
       disposer: () => {}, 
       next() {
          return lastPromise = lastPromise.then(() => new Promise(resolve => {
               resolver = resolve
          }))
       },
       return() { 
         disposer();
         return ({value: undefined, done: true });
       }
   }
})
```

Maybe missing out on values is confusing, and queuing is indeed a lot better. At least the semantics would be more clear. 

```javascript
ObservableValue.prototype.[Symbol.asyncIterator] = function() {
   let lastPromise = Promise.resolve()
   let resolver
   const queue = [this.get()]
   let disposer = this.observe(v => {  // observe is more efficient then reaction, but further the same
       if (resolver) resolver(v)
       else queue.push(v)
       resolver = null 
   }) 
   return {
       disposer: () => {}, 
       next() {
          return lastPromise = lastPromise.then(() =>
               queue.length
               ? Promise.resolve(queue.shift())
               : new Promise(resolve => {
                  resolver = resolve
               }
          )
       },
       return() { 
         disposer();
         return ({value: undefined, done: true });
       }
   }
})
```

**Comment by benjamingr** [comment]
Here is an implementation that yields all the values by accounting for both cases of `.next()` called several times before values are available or multiple values pushed before `.next()` is called:

```js
ObservableValue.prototype.[Symbol.asyncIterator] = function() {
  const producerQueue = []; // stuff that's generated but not consumed yet
  let consumerQueue = []; // stuff that's consumed but not generated yet
  const disposer = reaction(
    () => this.get(), 
    (value) => {
      if (consumerQueue.length > 0) {
        // there is a pending promise
        let resolver = consumerQueue.shift();
        resolver({value, done: false});
      } else {
        producerQueue.push(value);
      }
    }
  );
  return {
      async next() {
        if (producerQueue.length > 0) { // there is a pending value 
          return {value: producerQueue.shift(), done: false};
        }
        return await new Promise((resolve) => consumerQueue.push(resolve)); 
      },
      async return() { 
        disposer(); // maybe also invalidate all the consumerQueue? 
        return {value: undefined, done: true };
      }
  }
});
```

**Comment by benjamingr** [comment]
That's another interesting question, if we should immediately yield the original value or skip it

**Comment by mweststrate** [comment]
Yes, that was what I was doubting about as well. I think we probably should, as having users skip the first one if not interested is easy

**Comment by benjamingr** [comment]
Well, thinking of use cases, I think that producing observables from `asyncIterator`s is equally important - namely having `@observable` check if it's passed an objet with a `Symbol.asyncIterator` or an instance of `AsyncGeneratorFunction` and if it is, to `for...await` it and update a backing atom, that would enable the code from https://github.com/mobxjs/mobx/issues/1370#issuecomment-371194286

**Comment by mweststrate** [comment]
Is this a good test scenario:

```javascript
class Profile {
   name = observable.box()

   @generator *profile() {
        for await(const name of this.name) {
              await delay(500) // debounce
              // Q: how to skip to latest value here?
               yield await Promise.race([
                     myAsyncDataFetcher(name),
                     delay(10000).then(() =>  "timeout while fetchingdata")
              ])
        }
   }
}
```

**Comment by spion** [comment]
good catch there. debounce is difficult even with async generators.

We don't need the queueing or buffering for observables. If you request the next value 5 times in parallel, they should all get the (same) next value.

You will probably need `name = peek(this.name)` after delay

**Comment by benjamingr** [comment]
Back from lunch!

@mweststrate by consuming the iterator with `for...await` you are _by definition_ not skipping to the last value, if you want to debounce an observable I'm not sure I'd implement it this way (can implement the interface like we did for the conversion) - if we only emit the "latest" values then this isn't a problem (although it's a very specific case IMO).

(Also, worth mentioning that while promise cancellation isn't specced, `fetch` is cancellable with `AbortController` and XHR with `abort`, so a more "correct" implementation would deal with it - maybe) 

This is also because iterators are lazy and pull based whereas observables are eager and push based by design as a model. In this case the _easiest_ fix would be to go through both observables and then async iterators (observables are `Symbol.asyncIterator`).

**Comment by stephen-dahl** [comment]
almost afraid to post but there was so much great discussion happening on this till you it all of a sudden stopped back in March.
Did we come to an implementation that everyone agrees on?

**Comment by benjamingr** [comment]
> almost afraid to post but there was so much great discussion happening on this till you it all of a sudden stopped back in March.

Hey, please feel free to post. I know it can seem like _much_ but discussion and feedback is welcome and appreciated.

I'm not aware of anyone working on this and Michel is currently traveling - a PR or proposal is welcome :)

**Comment by ItamarShDev** [comment]
Wow, this an interesting thread!
Read it all in one straight shot. and will re-read again at least twice.
Very eye opening to me.

Though, to be honest, as an "average user", it seems like it could cause confusion.
Whenever we had to deal in such case at my team, we usually had either used `mobx-utils.fromPromise` as shown here, or changed the architecture of the component tree because it wasnt the best.

What I'm trying to day here is, that sometime simplicity needs to be considered with a bigger weight than usual.
Sometimes, the users learn and change bad code because the API is so simple, it forces the user to think (Rich Hickey's idea, not mine).

Just my two cents.
Now for second reading :)

**Comment by mweststrate** [comment]
Closing, I think it is a coolish idea, but in the end not *that* valuable.
